src/serac/numerics/functional/tensor.hpp

```diff
@@ -1170,58 +1172,147 @@ SERAC_HOST_DEVICE bool is_symmetric_and_positive_definite(tensor<double, 3, 3> A
 }
 
 /**
- * @brief Solves Ax = b for x using Gaussian elimination with partial pivoting
- * @param[in] A The coefficient matrix A
- * @param[in] b The righthand side vector b
- * @note @a A and @a b are by-value as they are mutated as part of the elimination
+ * @brief Compute LU factorization of a matrix with partial pivoting
+ * @param[in] A The matrix to factorize
+ * @return x tuple of [P, L, U], with
+ * L the lower triangular factor
+ * U the upper triangular factor
+ * P an array indicating the row permutations due to pivoting.
+ * Row i of A was exchanged with row P[i].
  */
 template <typename T, int n>
-SERAC_HOST_DEVICE constexpr tensor<T, n> linear_solve(tensor<T, n, n> A, const tensor<T, n> b)
+SERAC_HOST_DEVICE constexpr auto lu(const tensor<T, n, n>& A)
 {
+  // BT question for Sam Mish: Why not use the standard library abs?
   constexpr auto abs  = [](double x) { return (x < 0) ? -x : x; };
   constexpr auto swap = [](auto& x, auto& y) {
     auto tmp = x;
     x        = y;
     y        = tmp;
   };
 
-  tensor<double, n> x{};
+  auto           U = A;
+  auto           L = DenseIdentity<n>();
```

Comment:
Oh, ok - not good. I could use `make_tensor()` to make an identity with type `T`.  Is that a good fix?

Reply:
Fixed this in b0570fe2cd38e112980fdcff2d14622bd8740021 (with your coaching). I think the right thing to do long-term is to parameterize the identity matrix data type, so that you can write `L = Identity<T, n>();` here. I made a note of this in the comments. Since that's a breaking change, I'll defer that for now.